Refactor riak_api_pb_service to use callback mod attributes

rzezeski opened the pull request:
Use `-callback` module attributes instead of `behavior_info` to declare the callbacks. This plays better with dialyzer and will get rid of all the `Callback info about the riak_api_pb_service behaviour is not available` warnings across the projects.

Also updated the `decode` callback attribute and edoc to include the optional security namespace.

@Vagabond @seancribbs Might want to give a quick look around my "security namespace" addition. I'm not 100% that is the best name or if I documented it properly.

rzezeski committed:
Use -callback specs instead of behavior_info

As of Erlang 15B a new callback spec was added which allows callbacks to
be precisely defined as well as work with dialyzer.

rzezeski committed:
Update edoc for decode callback

Add optional security namespace to decode callback edoc and update
encode spec.

rzezeski commented:
Here is diff of dialyzer output showing removal of `not available`
errors.

```
# diff -u d.txt  d7.txt | egrep '^-'
-riak_api_basic_pb_service.erl:45: Callback info about the riak_api_pb_service behaviour is not available
-riak_core_pb_bucket.erl:45: Callback info about the riak_api_pb_service behaviour is not available
-riak_core_pb_bucket_type.erl:35: Callback info about the riak_api_pb_service behaviour is not available
```

seancribbs commented on src/riak_api_pb_service.erl:
```diff
@@ -174,16 +175,33 @@
-    undefined.
+-callback init() -> State :: term().
+
+-type security_namespace() :: {Level1 :: term(), Level2 :: term()}.
```
This type should probably be in `riak_core_security` and the actual type should probably be `{Permission :: string(), Target :: term()}`. I just noticed we don't have any type annotations in `riak_core_security`. :frowning:

rzezeski replied:
Okay, I think that makes sense. I'll wait to see if @Vagabond weighs in on the naming and types but I like perm/target. +1 to putting the type where it belongs. I can do that too once Andrew weighs in.

rzezeski commented:
I fixed the first commit, rebased, and forced pushed. Rather than get a type into the security module I'm adding one locally as a stand in for now so this PR doesn't bit rot.

Once again, removes the following errors from dialyzer:

```
# diff -u d.txt d9.txt | grep '^-riak_'
-riak_api_basic_pb_service.erl:45: Callback info about the riak_api_pb_service behaviour is not available
-riak_core_pb_bucket.erl:45: Callback info about the riak_api_pb_service behaviour is not available
-riak_core_pb_bucket_type.erl:35: Callback info about the riak_api_pb_service behaviour is not available
```

All tests pass:

```
=======================================================
  All 32 tests passed.
Cover analysis: /root/work/riak_api/.eunit/index.html
```

seancribbs commented on src/riak_api_pb_service.erl:
```diff
@@ -63,19 +63,23 @@
 %%     Code = non_neg_integer()
 %%     Message = binary()
 %%     DecodedMessage = Reason = term()
+%%     SecurityNamespace = {term(), term()}
```
Can you sync this up with the type you defined below?

rzezeski commented:
Check out the edoc now. Think I got it all.

seancribbs commented:
@rzezeski Yeah I guess I was wondering why define the type later in the file and then not name it specifically in the spec for that callback. See https://github.com/basho/riak_api/pull/53/files#diff-a918ef5ee14869298c376b8ec15bcbe9R180

seancribbs commented:
Sorry, I meant the doc, not the spec. :(

rzezeski committed:
Use already defined type in the edoc

rzezeski commented:
What about now?

seancribbs commented:
Thanks @rzezeski. Nitpicks over, entering testing.

seancribbs commented:
:+1: to merge
